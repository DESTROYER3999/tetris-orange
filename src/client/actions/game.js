import socket from "../services/socket-api";
import {
  KEY_PRESSED,
  UPDATE_SPECTRES,
  UPDATE_GAME_MODE
} from "../../constants/actionTypes";

export const handleKeyPress = e => {
  socket.emit(KEY_PRESSED, { code: e.code });
};

export const updateSpectres = spectres => ({
  type: UPDATE_SPECTRES,
  spectres
});

export const updateGameMode = mode => {
  return () => {
    socket.emit(UPDATE_GAME_MODE, mode);
  };
};
