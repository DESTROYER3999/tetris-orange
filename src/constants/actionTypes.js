export const JOIN_ROOM = "JOIN_ROOM";
export const UPDATE_ROOM = "UPDATE_ROOM";
export const LEAVE_ROOM = "LEAVE_ROOM";
export const LOBBY_ROOM = "LOBBY_ROOM";

export const NEW_PLAYER = "NEW_PLAYER";
export const DISCONNECT_PLAYER = "DISCONNECT_PLAYER";
export const UPDATE_PLAYER = "UPDATE_PLAYER";
export const UPDATE_PLAYERS_LIST = "UPDATE_PLAYERS_LIST";
export const DELETE_PLAYER = "DELETE_PLAYER";

export const HANDLE_HASH = "HANDLE_HASH";

export const UPDATE_BOARD = "UPDATE_BOARD";
export const UPDATE_GAME = "UPDATE_GAME";
export const UPDATE_GAME_MODE = "UPDATE_GAME_MODE";
export const UPDATE_SPECTRES = "UPDATE_SPECTRES";

export const DISPLAY_TOAST = "DISPLAY_TOAST";

export const CONNECTION = "connection";
export const DISCONNECT = "disconnect";

export const ADD_CHAT_MESSAGE = "ADD_CHAT_MESSAGE";
export const NEW_CHAT_MESSAGE = "NEW_CHAT_MESSAGE";

export const KEY_PRESSED = "KEY_PRESSED";
