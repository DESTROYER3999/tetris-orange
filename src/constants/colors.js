export const MALUS = "#3d3d3d";

export const VIVID_BLUE = "#143cdc";

export const CYAN_BLUE = "#14d4dc";

export const CRIMSON = "#dc143c";

export const PURE_ORANGE = "#ff8b00";

export const LIME_GREEN = "#3cdc14";

export const PURE_VIOLET = "#b600ff";

export const PURE_YELLOW = "#fff400";

export const DARK_CYAN = "#375f64";

export const GREEN_CHAT = "#73F054";

export const RED_CHAT = "#ff3366";
