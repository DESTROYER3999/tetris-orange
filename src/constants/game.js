export const SOLO = "Solo";
export const BATTLEROYAL = "Battleroyal";
export const MAX_PLAYER_SOLO = 1;
export const MAX_PLAYER_BATTLEROYAL = 5;
export const BOARD_HEIGHT = 20;
export const BOARD_WIDTH = 10;
