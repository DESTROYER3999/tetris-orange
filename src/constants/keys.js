export const enumKeys = {
  ARROW_LEFT: "ArrowLeft",
  ARROW_RIGHT: "ArrowRight",
  ARROW_DOWN: "ArrowDown",
  ARROW_UP: "ArrowUp",
  SPACE: "Space",
  ENTER: "Enter"
};
