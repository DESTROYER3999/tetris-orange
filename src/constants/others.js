export const LOG_LINE =
  "\n_________________________________________________________________________________________________________\n\n";
