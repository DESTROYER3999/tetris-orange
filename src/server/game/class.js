import { updateHighScoresInDb } from "../../database";
class Game {
  constructor() {
    this._rooms = [];
    this._players = [];
    this._highscores = [
      {
        score: 0,
        name: ""
      },
      {
        score: 0,
        name: ""
      },
      {
        score: 0,
        name: ""
      },
      {
        score: 0,
        name: ""
      },
      {
        score: 0,
        name: ""
      }
    ];
  }

  get rooms() {
    return this._rooms;
  }

  get players() {
    return this._players;
  }
  get highscores() {
    return this._highscores;
  }

  set highscores(highscores) {
    this._highscores = highscores;
  }

  addRoom(room) {
    this._rooms.push(room);
  }

  removeRoom(roomName) {
    this._rooms = this._rooms.filter(room => room.name !== roomName);
  }

  addPlayer(player) {
    this._players.push(player);
    return player;
  }

  removePlayer(playerId) {
    this._players = this._players.filter(player => player.id !== playerId);
  }

  findRoom(roomName) {
    return this._rooms.find(room => room.name === roomName);
  }

  findPlayer(playerId) {
    return this._players.find(player => player.id === playerId);
  }

  updateHighScores(score, playerName) {
    const newHighScore = { score, playerName };
    this._highscores.every((highscore, index) => {
      if (highscore.score < score) {
        this._highscores.splice(index, 0, newHighScore);
        this._highscores.length = 5;
        updateHighScoresInDb(this._highscores);
        return false;
      } else return true;
    });
  }

  createPublicRoomsArray() {
    return this._rooms.map(room => ({
      name: room.name,
      playersCount: room.playersCount,
      mode: room.mode,
      isStarted: room.isStarted
    }));
  }

  toObject() {
    return {
      rooms: this.createPublicRoomsArray(),
      highscores: this._highscores
    };
  }
}

const instance = new Game();

export default instance;
