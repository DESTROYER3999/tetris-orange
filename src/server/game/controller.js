import Game from "./class";
import Timer from "../timer/class";
import { UPDATE_GAME, DISPLAY_TOAST } from "../../constants/actionTypes";
import { BATTLEROYAL } from "../../constants/game";
import { updateRoomClientSide } from "../room/controller";
import { getHighScoresFromDb } from "../../database";

export const initClientState = async io => {
  await getHighScoresFromDb();
  updateGameClientSide(io);
};

const handleInterval = (room, io) => {
  room.players.forEach(player => {
    if (player.inGame) {
      if (!player.piece.moveDown(player.heap)) {
        player.updateHeap();
      }
    }
  });
  updateRoomClientSide(room, io);
};

export const startGame = (room, io, socket) => {
  if (room.isStarted) return;
  if (room.mode === BATTLEROYAL && room.playersCount <= 1) {
    socket.emit(DISPLAY_TOAST, {
      type: "error",
      message: "Not enough players"
    });
    return;
  }
  room.newGame();
  updateRoomClientSide(room, io);
  room.timer = new Timer(() => handleInterval(room, io), room.speed);
  room.timer.start();
};

export const updateGameClientSide = io => {
  io.emit(UPDATE_GAME, { game: Game.toObject() });
};
