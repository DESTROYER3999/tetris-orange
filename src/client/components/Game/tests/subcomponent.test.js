import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import GameSub from "../subcomponent";
import socket from "../../../services/socket-api";
import { SOLO, BATTLEROYAL } from "../../../../constants/game";

beforeEach(done => {
  // Setup
  if (socket.disconnected) {
    socket.connect();
  }
  socket.on("connect", () => {
    done();
  });
});

afterEach(done => {
  //   Cleanup;
  if (socket.connected) {
    socket.disconnect();
  }
  done();
});

describe("<GameSub />", () => {
  test("GameSub rendering in solo mode", () => {
    const wrapper = shallow(<GameSub mode={SOLO} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
  test("GameSub rendering in battleroyal mode", () => {
    const wrapper = shallow(<GameSub mode={BATTLEROYAL} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
