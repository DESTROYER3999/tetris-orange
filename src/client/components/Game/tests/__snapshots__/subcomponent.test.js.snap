// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameSub /> GameSub rendering in battleroyal mode 1`] = `
<styled.div
  id="game"
>
  <styled.div
    id="left-item"
  >
    <Connect(SpectreSub) />
    <Connect(BoardSub) />
    <styled.div>
      <Connect(NextPieces) />
    </styled.div>
  </styled.div>
  <styled.div
    id="right-item"
  >
    <Connect(GameInfoSub) />
    <Connect(ChatSub) />
  </styled.div>
</styled.div>
`;

exports[`<GameSub /> GameSub rendering in solo mode 1`] = `
<styled.div
  id="game"
>
  <styled.div
    id="left-item"
  >
    <Connect(SpectreSub) />
    <Connect(BoardSub) />
    <styled.div>
      <Connect(NextPieces) />
      <Connect(ProgressionSub) />
    </styled.div>
  </styled.div>
  <styled.div
    id="right-item"
  >
    <Connect(GameInfoSub) />
    <Connect(ChatSub) />
  </styled.div>
</styled.div>
`;
